Test TF-M config with p256-m driver

mpg opened the pull request:
## Description

Resolves #8006.

Note: will need rebasing if #8203 gets merged first.

~Built on top of~ Supersedes #8032 ~itself built on top of #7992 (merged)~.

One problem that was uncovered by #8032 is that the p256-m driver does not provide all entry points we'd want in order to be able to get rid of `ECP_C` and `BIGNUM_C`, namely it's missing the `import_key` and `export_public_key` entry points. This PR solve that:
- Before adding the entry points, we need to expand the public API in `p256-m.h` a bit.
- Then we add the entry points.
- Then add a JSON file, as unlike the entry points that were introduced before, those have programmatic generation of the driver wrappers.
- Then a few extra clean-ups in the p256-m driver while at it.

## PR checklist

- [x] **changelog** deferred, see #7757
- [x] **backport** not required - new feature
- [x] **tests** provided

mpg commented:
Removing "needs-preceding-pr" as #7992 has been merged now, and I'm closing #8032 in favour of this one - Valerio is away so it makes more sense for me to continue here and us to have only one PR to manage.

I'll rebase to resolve conflicts in a moment, before looking at pending comments.

mpg commented:
I rebased over the last state of #8032, rather than on development, in order to make sure we're not losing anything by closing 8032 in favour of this one. This was enough to resolve conflicts too.

mpg commented:
I went of the the unresolved comments in #8032 and double-checks they had all been addressed (one of them by an independent PR) before marking the conversations as resolved.

mpg commented on scripts/data_files/driver_jsons/p256_transparent_driver.json:
```diff
@@ -0,0 +1,19 @@
+    "capabilities": [
+        {
+            "mbedtls/c_condition": "defined(MBEDTLS_P256M_EXAMPLE_DRIVER_ENABLED)",
+            "entry_points": ["import_key", "export_public_key"],
```
Addressed in https://github.com/Mbed-TLS/mbedtls/pull/8041/files#diff-e683396c93d3af1ea2db5a3d94788da3916ed084f3eb8be5e5f7c0e429badf39R2820

mpg commented:
I think this is now ready for review.

gilles-peskine-arm commented:
I think we should rename the option `MBEDTLS_P256M_EXAMPLE_DRIVER_ENABLED` since we intend to release it as fit for production, not as an example. Since this will conflict with the changes here to test it, and this pull request is ready for review, I propose to do the renaming in a follow-up.

mpg commented:
Note that the present PR is likely to be rebased if #8164 gets merged first (which seems likely considering 8164 already has 1 approval and is smaller than the present PR which has 0 complete review so far). So if you feel like doing a renaming PR quickly, I can probably rebase this on both 8164 and the renaming PR at once.

tom-daubney-arm reviewed (commented):
I have not done a full review but I was noticing some nits in the written English so I wanted to collate them and submit them now.

xkqian commented on include/mbedtls/pk.h:
```diff
@@ -222,6 +222,28 @@ typedef struct mbedtls_pk_rsassa_pss_options {
 
+/* Internal helper to define which fields in the pk_context structure below
+ * should be used for EC keys: legacy ecp_keypair or the raw (PSA friendly)
+ * format. It should be noticed that this only affect how data is stored, not
```
```suggestion
 * format. It should be noticed that this only affects how data is stored, not
```

xkqian commented on include/mbedtls/pk.h:
```diff
@@ -222,6 +222,28 @@ typedef struct mbedtls_pk_rsassa_pss_options {
+ * format. It should be noticed that this only affect how data is stored, not
+ * which functions are used for various operations. The overall picture looks
+ * like this:
+ * - if USE_PSA is not defined and ECP_C is then use ecp_keypair data structure
```
```suggestion
 * - if USE_PSA is not defined and ECP_C is defined then use ecp_keypair data structure
```

xkqian commented on include/mbedtls/pk.h:
```diff
@@ -222,6 +222,28 @@ typedef struct mbedtls_pk_rsassa_pss_options {
+#if defined(MBEDTLS_USE_PSA_CRYPTO) && defined(PSA_WANT_KEY_TYPE_ECC_PUBLIC_KEY) && \
+    !defined(MBEDTLS_ECP_C)
+#define MBEDTLS_PK_USE_PSA_EC_DATA
+#endif /* MBEDTLS_USE_PSA_CRYPTO && !MBEDTLS_ECP_C */
```
```suggestion
#endif /* MBEDTLS_USE_PSA_CRYPTO &&
                PSA_WANT_KEY_TYPE_ECC_PUBLIC_KEY &&
                !MBEDTLS_ECP_C */
```

mpg replied:
Good point. I just removed the comment, as the corresponding `#if` is only 2 lines above, with no nesting, so I don't think the comment was useful. (See [our coding standard](https://mbed-tls.readthedocs.io/en/latest/kb/development/mbedtls-coding-standards/#preprocessor-directives): "Add a comment to the `#endif` directive if the distance to the opening directive is bigger than a few lines or contains other directives".)

xkqian commented on tests/scripts/all.sh:
```diff
@@ -2780,6 +2780,106 @@ component_test_psa_crypto_config_reference_ecc_no_bignum () {
+    common_tfm_config
+
+    # Set the list of accelerated components in order to remove them from
+    # builtin support. We don't set IMPORT and EXPORT because P256M does not
```
I am confused, we added p256_transparent_import_key() and p256_transparent_export_public_key() in this PR, why we declare that P256 doesn't support the operations?

mpg replied:
Ah, good point, I forgot to update the comment. Also, we _do_ set IMPORT and EXPORT right below the comment, so we had a case of comment contradicting the code it applies to... thanks for catching this!

xkqian reviewed (commented):
Just some nitpick improvements and a small question.

tom-daubney-arm commented:
I know this isn't part of the PR but I noticed a spelling mistake while I was rooting around in `p246-m.c` on line `1009`. It currently reads `* out: x, y in Mongomery domain` but should read `* out: x, y in Montgomery domain` (missing "t"). I can't do a suggestion on github since it is not part of the PR (at least I don't think so).
If there is a better place to fix this please let me know and I will comment there.

mpg commented:
> I know this isn't part of the PR but I noticed a spelling mistake while I was rooting around in `p246-m.c` on line `1009`. It currently reads `* out: x, y in Mongomery domain` but should read `* out: x, y in Montgomery domain` (missing "t"). I can't do a suggestion on github since it is not part of the PR (at least I don't think so).
> If there is a better place to fix this please let me know and I will comment there.

I've added it to my notes. Gilles has noticed a few other issue with the comment, I'll fix all of them at once at some point. Thanks!

mpg commented on 3rdparty/p256-m/p256-m_driver_entrypoints.c:
```diff
@@ -59,6 +70,82 @@ static psa_status_t p256_to_psa_error(int ret)
+        return PSA_ERROR_NOT_SUPPORTED;
+    }
+
+    /* Validate the key (and its type and size) */
```
I think we're doing more that just validating the type and size: we are indeed validating the key as well, so I think the current comment is correct.

mpg commented:
Thanks for your (partial) reviews! I've addressed your points (except on where I explained one). I've also noticed while fixing typos that one block of comments+define somehow ended up triplicated, so I removed the extra copies.

This is now ready for re-review / completing the review in progress.

paul-elliott-arm commented on 3rdparty/p256-m/p256-m_driver_entrypoints.c:
```diff
@@ -149,21 +229,23 @@ psa_status_t p256_transparent_sign_hash(
+        return PSA_ERROR_CORRUPTION_DETECTED;
+    }
+    if (signature_size < SIGNATURE_SIZE) {
+        return PSA_ERROR_BUFFER_TOO_SMALL;
```
Documentation for this function does not contain all error codes:

`PSA_ERROR_BUFFER_TOO_SMALL` and `PSA_ERROR_CORRUPTION_DETECTED` are missing (along with anything that `p256_ecdsa_sign()` can return), `PSA_ERROR_NOT_SUPPORTED` needs removing.

mpg commented on 3rdparty/p256-m/p256-m_driver_entrypoints.c:
```diff
@@ -149,21 +229,23 @@ psa_status_t p256_transparent_sign_hash(
-        return status;
+    /* Validate sizes, as p256-m expects fixed-size buffers */
+    if (key_buffer_size != PRIVKEY_SIZE) {
+        return PSA_ERROR_CORRUPTION_DETECTED;
```
My reasoning is that (1) the key has been validated on import, and (2) driver dispatch should only call this driver for keys on the right curve. So, if this check fails, it means the core and/or dispatch is misbehaving.

We could almost skip the checks and just trust the core (this is a transparent driver, so there's no security barrier between the core and the driver) , but since this function wraps an API with "no explicit length because it's 32 all the time" into an interface with "explicit length parameter", I felt it was cleaner for the wrapper to assert that the length is as expected.

mpg replied:
I created a straw poll on slack to see what others think.

tom-daubney-arm replied:
I was going to ask a bit about this. I feel like `_CORRUPTION_DETECTED` is maybe too precise for a situation that may not be caused by corruption. This is a bit hand-wavey of me but I feel like we could get an incorrectly sized key for multiple reasons (a bug elsewhere we have yet to find or some part of the integration going wrong or indeed corruption) and so `_INVALID_ARGUMENT` could be a better suited catch-all error to report.

mpg replied:
Ok, I'll change those to `INVALID_ARGUMENT` then!

tom-daubney-arm replied:
Cool, ok thanks!

gilles-peskine-arm replied:
Indeed, thinking about this a bit more I think `CORRUPTION_DETECTED` should be reserved for cases that are impossible, short of physical corruption during execution or a major bug in the code. We shouldn't use this for the kind of mistakes that can happen during configuration or integration, such as calling a driver on a key type that it doesn't support.

`CORRUPTION_DETECTED` should only catch bugs within code that's maintained by the same entity. The code path here involves the PSA core, a plug-in which is maintained in the same repository but in a separate directory under the heading “third-party”, possibly in the future a plug-in related to key parsing, a driver description file which may have been modified by a distributor or integrator with only vague knowledge of the various components involved, and a driver description generator/compiler script. With that many entities, if an error happens, we should convey “integration mistake”, not “memory corruption”.

Here either `NOT_SUPPORTED` (“I don't understand this, I guess it's a curve I don't support”) or `INVALID_ARGUMENT` (“I don't understand this, I have no idea what you mean”) seem appropriate.

tom-daubney-arm reviewed (commented):
Just dropping a comment to say I have reviewed and this mostly looks great, except for the few things that have already been mentioned. I have nothing further to add right now. Will re-review as soon as it's ready.

mpg commented:
Thanks @paul-elliott-arm and @tom-daubney-arm for your reviews! I believe I have addressed your feedback, so this is ready for re-review.

paul-elliott-arm commented:
> Aw, CI came back red, failing with unexpected discrepancies in "driver test_tfm_config_p256m_driver_accel_ec vs reference test_tfm_config" which is exactly what matters in this PR, so does not look spurious.
> 
> The CI was green 3 commits ago and I really didn't expect any of the last 3 commits to break it like this, but here we are... I'll investigate, fix, and let you know when this is ready to re-review.

Looks like a single missing '*' from p256-m_driver_entrypoints.c, line 206 - do you want me to fix this before tomorrow?

mpg commented:
Ah, right, I was looking at "result analysis" while I should have been looking at all.sh components. I'm fixing this, but thanks for the offer! (Finished early yesterday due to an appointment, making up for it now.)

mpg commented:
Fix pushed, should be ready for review again!

tom-daubney-arm reviewed (approved):
LGTM, thanks for the changes!
